ksql-parser/src/main/java/io/confluent/ksql/parser/tree/PrimitiveType.java

```diff
@@ -15,51 +15,142 @@
+      .put(SqlType.BIGINT,  p -> new PrimitiveType(SqlType.BIGINT))
+      .put(SqlType.DOUBLE,  p -> new PrimitiveType(SqlType.DOUBLE))
+      .put(SqlType.STRING,  p -> new PrimitiveType(SqlType.STRING))
+      .put(SqlType.DECIMAL, p -> new PrimitiveType(SqlType.DECIMAL, p))
+      .build();
 
   public static PrimitiveType of(final String typeName) {
+    return of(typeName, ImmutableList.of());
+  }
+
+  public static PrimitiveType of(
+      final String typeName,
+      @Nullable final List<Integer> typeParameters
+  ) {
     switch (typeName.toUpperCase()) {
       case "INT":
         return PrimitiveType.of(SqlType.INTEGER);
       case "VARCHAR":
         return PrimitiveType.of(SqlType.STRING);
+      case "DEC": // SQL standard states that DEC is similar to DECIMAL
+        return PrimitiveType.of(SqlType.DECIMAL, typeParameters);
       default:
         try {
           final SqlType sqlType = SqlType.valueOf(typeName.toUpperCase());
-          return PrimitiveType.of(sqlType);
+          return PrimitiveType.of(sqlType, typeParameters);
         } catch (final IllegalArgumentException e) {
           throw new KsqlException("Unknown primitive type: " + typeName, e);
         }
     }
   }
 
   public static PrimitiveType of(final SqlType sqlType) {
-    final PrimitiveType primitive = TYPES.get(Objects.requireNonNull(sqlType, "sqlType"));
-    if (primitive == null) {
+    return function(sqlType).apply(ImmutableList.of());
+  }
+
+  public static PrimitiveType of(
+      final SqlType sqlType,
+      @Nullable final List<Integer> sqlTypeParameters
+  ) {
+    if (sqlTypeParameters == null) {
+      return function(sqlType).apply(ImmutableList.of());
+    } else {
+      return function(sqlType).apply(ImmutableList.copyOf(sqlTypeParameters));
+    }
+  }
+
+  private static Function<ImmutableList<Integer>, PrimitiveType> function(final SqlType sqlType) {
+    final Function<ImmutableList<Integer>, PrimitiveType> function =
+        TYPES.get(Objects.requireNonNull(sqlType, "sqlType"));
+    if (function == null) {
       throw new KsqlException("Invalid primitive type: " + sqlType);
     }
-    return primitive;
+
+    return function;
   }
 
   private PrimitiveType(final SqlType sqlType) {
+    this(sqlType, ImmutableList.of());
+  }
+
+  private PrimitiveType(final SqlType sqlType, final ImmutableList<Integer> typeParameters) {
     super(Optional.empty(), sqlType);
+    this.typeParameters = typeParameters;
+
+    // Verify the passed parameters are correct for the specified schema
+    checkTypeParameters();
+  }
+
+  public ImmutableList<Integer> getSqlTypeParameters() {
+    return typeParameters;
+  }
+
+  private void checkTypeParameters() {
+    switch (getSqlType()) {
+      case DECIMAL:
+        checkDecimalTypeParameters();
+        break;
+      default:
+        if (typeParameters.size() > 0) {
+          throw new KsqlException(
+              "Primitive type does not support parameters: " + getSqlType());
```

Comment:
Change the error message to indicate the type is not supported since you are showing the type name not the parameters!

Reply:
Done